### PR TITLE
fix: treat 2xx token responses carrying an OAuth error as refresh failures

### PR DESCRIPTION
**Pull request by daviddanialy**
AIM-201

## Summary

- `refreshSessionTokens` now parses the body of a 2xx token response that carries no `access_token`. If it holds an OAuth error, the refresh fails with a `TokenRefreshError` that carries the error code and the real HTTP status, so `refreshOutcomeForError` files it as an upstream rejection and an `invalid_grant` clears the stored refresh grant exactly as a 4xx would. A 2xx body with neither a token nor an error is unchanged and stays transient.
- `oautherr.CanonicalTokenErrorCode` maps GitHub's `bad_refresh_token` and `bad_verification_code` onto `invalid_grant` by exact match. `ParseTokenError` still returns extension codes verbatim; both `TokenRefreshError` constructors canonicalise after parsing and keep the original in `UpstreamCode()`.
- Both refresh-failure log lines (request path and scheduled keepalive) record the upstream's error code as `gram.oauth.error`. The success-body cause names only the code, never the body, since a success-status response can still carry token material.
- Tests: canonicalisation, three new outcome-classifier rows, and two end-to-end `RefreshNow` cases against a fake token endpoint (200 with `bad_refresh_token` clears the grant; empty 200 body keeps it).

## Motivation

GitHub's token endpoint answers a revoked or expired refresh token with HTTP 200 and `{"error":"bad_refresh_token"}`. The refresh path only classified non-2xx bodies, so this became a code-less "returned no access token" error: never `invalid_grant`, refresh token never cleared, and the same dead grant retried on every proxied request and every keepalive sweep, while the session kept reading as connected. Over the last 30 days this produced roughly 1,500 failed refreshes from a single connection.

The authorization-code exchange has the same blind spot; it does not parse error bodies yet, so the `bad_verification_code` alias only takes effect once it does. That is left as a follow-up because no session exists there to clear, only the error message improves.

Temporal actions/month ≈ 0, scales with fixed (no workflow, activity, schedule, or signal is added or changed; the keepalive activity only gains a log attribute).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01C555wgLmfbVWf4oxK7FvwB

### CODE REVIEW

File: server/internal/oautherr/vendor_error.go [hunk 1]
```diff
@@ -102,6 +103,26 @@ func (v vendorCodeMessage) normalized() vendorCodeMessage {
 	}
 }
 
+// vendorDeadGrantCodes maps RFC-shaped extension codes that a provider sends
+// only when a refresh grant can never succeed again onto CodeInvalidGrant.
+// GitHub's token endpoint (https://github.com/login/oauth/access_token)
+// answers a revoked or expired refresh token with "bad_refresh_token" under
+// HTTP 200. Matching is exact: any other extension code is returned as is.
+var vendorDeadGrantCodes = map[string]string{
+	"bad_refresh_token": CodeInvalidGrant,
```
[the remark this answers is not in the document]
Comment (daviddanialy) [reply server/internal/oautherr/vendor_error.go]: Added in 3313a8c132: `bad_verification_code` now maps to `invalid_grant` alongside `bad_refresh_token`, with the table comment and `TestCanonicalTokenErrorCode` updated. The authorization-code exchange itself still does not call `CanonicalTokenErrorCode`, as the PR body notes.

File: server/internal/remotesessions/refreshservice_invalid_grant_test.go [hunk 2]
```diff
@@ -119,6 +119,44 @@ func TestRefreshNow_InvalidGrant_DubRefreshTokenNotFound_ClearsRefreshGrant(t *t
...
+	t.Parallel()
+
+	active, tokenErr := refreshNowAgainstUpstreamError(
+		t,
+		"refreshnow-github-200-error",
+		http.StatusOK,
+		`{"error":"bad_refresh_token","error_description":"The refresh token passed is incorrect or expired.","error_uri":"https://docs.github.com/apps/oauth-apps/troubleshooting"}`,
+	)
+
+	require.Equal(t, "invalid_grant: The refresh token passed is incorrect or expired.", tokenErr.Reason)
```
[the remark this answers is not in the document]
Comment (daviddanialy) [reply server/internal/remotesessions/refreshservice_invalid_grant_test.go:140]: Done in 3313a8c132: `refreshNowAgainstUpstreamError` now also returns the `*RefreshError`, and each test asserts its `Outcome` (GitHub 200 + `bad_refresh_token` → `invalid_grant`, empty 200 → `internal_error`, plus the existing Datadog/Dub cases). Dropping `statusCode` back to 0 now fails the GitHub test.

File: server/internal/remotesessions/tokenservice.go [hunk 3]
```diff
@@ -495,21 +495,28 @@ func (m *ChallengeManager) validateAndRefresh(
...
+// upstream-refresh metric recorded for the attempt, so the log line joins to
+// its series; the user id is what lets "how many users are affected" be
+// answered, since a client id is one row per provider connection, not per
+// user.
 func refreshFailureAttrs(sess remotesessions_repo.RemoteSession, err error) []any {
 	reason := "upstream token refresh failed"
+	code := ""
 	if refreshErr, ok := errors.AsType[*TokenRefreshError](err); ok {
 		reason = refreshErr.Reason
+		code = refreshErr.UpstreamCode()
```
[the remark this answers is not in the document]
Comment (daviddanialy) [reply server/internal/remotesessions/tokenservice.go:509]: Leaving as is. The canonical form is already on the same log line as `gram.outcome=invalid_grant` (from `RefreshError.Outcome`), so `gram.oauth.error` logging `invalid_grant` too would be redundant and would drop the only record of what the upstream actually sent. Every other `gram.oauth.error` site logs the `error` parameter verbatim, and this keeps that meaning.

**Review by bflad — APPROVED** [review]
Looks good to me. Maybe someday vendors will implement standards across the board. 😅
